### PR TITLE
Fix "Memory Leak" when reverting to a Snapshot

### PR DESCRIPTION
**Pull request by ritzdorf**
### What I did

Currently, when `boa.env.anchor()` is used to revert to a previous snapshot, the evm state is reverted, but newly created contracts that have been registered since the snapshot has been taken, are not unregistered. Hence, repeated snapshots lead to an ever growing memory consumption. Hence, I added code to unregister contracts that have been deployed since the last snapshot has been taken.

### How I did it

I added a list to remember to newly deployed contracts as `_new_contracts`. The list elements are tuples. They contain:
- `addr`: the address, of type `Address`
- `novel_bytecode`: a boolean indicating whether or not the bytecode of this contract was seen for the first time as part of this deployment

The `_new_contracts` list is cleared when a snapshot is taken.

When the state is reverted to a previous snapshot, the newly created contracts are unregistered. Where the bytecode was novel, it is also unregistered.

### How to verify it

``` python
import boa
import psutil
import os

def get_process_memory():
    process = psutil.Process(os.getpid())
    return process.memory_info().rss / (1024 * 1024)  # Return memory usage in megabytes

novel_src = """

novel_value: public(uint256)
replace_me: public(uint256)

"""
known_src = """

known_value: public(uint256)

"""

known_contract = boa.loads(known_src)
known_contract.eval("self.known_value += 1")

for i in range(500):
    with boa.env.anchor():
        novel_contract = boa.loads(novel_src.replace("replace_me", f"var{i}"))
        novel_contract.novel_value()
        new_known_contract = boa.loads(known_src)
        new_known_contract.known_value()

print(known_contract.known_value())
print(f"Memory Usage: {get_process_memory()} MB")
```
This script with the current master:
```
1
Memory Usage: 212.2890625 MB
```

This script with the new version:
```
1
Memory Usage: 65.3125 MB
```

### Description for the changelog

Unregistering newly created contracts when reverting to snapshot

### Cute Animal Picture

Capybara

### CODE REVIEW

File: boa/environment.py [hunk 1]
```diff
@@ -117,7 +120,22 @@ def register_contract(self, address, obj):
...
+        novel_bytecode = bytecode not in self._code_registry
+        if novel_bytecode:
+            self._code_registry[bytecode] = obj
+        # Remember this address as a new contract
+        self._new_contracts.append((addr, novel_bytecode))
+
+    # Unregister a contract, e.g., while reverting to a snapshot
+    # addr should be an Address
+    def unregister_contract(self, addr, unregister_bytecode=True):
+        if addr.canonical_address in self._contracts:
```
Comment (charles-cooper) [inline boa/environment.py:132]: `self._contracts.pop(addr.canonical_address, None)` to save a lookup

File: boa/environment.py [hunk 2]
```diff
@@ -117,7 +120,22 @@ def register_contract(self, address, obj):
...
+            del self._contracts[addr.canonical_address]
+
+        if unregister_bytecode:
+            bytecode = self.evm.get_code(addr)
+            if bytecode in self._code_registry:
```
Comment (charles-cooper) [inline boa/environment.py:137]: ditto, `self._code_registry.pop(bytecode, None)`

File: boa/environment.py [hunk 3]
```diff
@@ -152,10 +170,15 @@ def reset_gas_used(self):
     @contextlib.contextmanager
     def anchor(self):
         snapshot_id = self.evm.snapshot()
+        # Remember new contracts from this snapshot
+        self._new_contracts = []
```
Comment (charles-cooper) [inline boa/environment.py:174]: i don't think this is correct in the presence of nested anchors, since the outer journal will get discarded

Comment (DanielSchiavini) [reply boa/environment.py:174]: I think we need an actual journal to encapsulate any data that changes in between snapshots

**Comment by ritzdorf** [comment]
Thanks for the feedback. I will try to get it to it. Am a bit busy though.

File: boa/environment.py (hunk 3 above)
Comment (charles-cooper) [reply boa/environment.py:174]: or maybe something really "clever", like a weakdict

**Comment by ritzdorf** [comment]
Hi @DanielSchiavini @charles-cooper,
thanks to @trocher we made the journaling changes here. We tested it and so far it works well for us.

**Pull request closed without merging**